In [107]:
import pandas as pd
import numpy as np

import scipy.stats as stats
import statsmodels.api as sm
from statsmodels.formula.api import ols
import matplotlib.pyplot as plt


In [108]:
# loading our dataframes
#we will be using df_m for our sample of math class studetns and df_p for our sameple of portugees class. 

df_m = pd.read_csv('../data/student-mat.csv', delimiter=';')
df_p = pd.read_csv('../data/student-por.csv', delimiter=';')

Before we start our statistical analysis, we need to determine 2 things, hypothesis (H1) and statisitcal test we can use to reject or fail to reject our null hypothesis (H0)

Hypothesis: 
    The main hypothesis is to see is there any effect of being in a relationship on the cummulative grade for the school. For male and female students. 

Therefore: 

    H0: There is no main effect of relationship status and gender on the cumulative grades of the students.
    H1: There is a main effect of relationship status and gender on the cumulative grades of the students. 

Note: 
    From our exploration stage, we identify 1 dependent variable (quantitative variable that we want to compare across different groups) 2 independet variables (categorical variables that define our groups)
        Dependet variable - commulative grades (num values, sum of G1 G2 and G3)
        Independet varaibles - Relationship status (singe or taken) and gender (male or female)

Statistical Test: 
    The perfect statistical test in our case is Two-way ANOVA

Our new Hypothesis are: 
    H0: There is no main effect of relationship status on the cumulative grades of the students.
    H1: There is a main effect of relationship status on the cumulative grades of the students.
    
    H0: There is no main effect of gender on the cumulative grades of the students.
    H1: There is a main effect of gender on the cumulative grades of the students.


In [109]:
#First we need to create the sum of grades across 3 semester for each class
sum_grades_m = df_m['G1'] + df_m['G2'] + df_m['G3']
sum_grades_p = df_p['G1'] + df_p['G2'] + df_p['G3']

In [110]:
#We need to add this column to our new DF for this test
test_df_m = df_m.assign(grades = sum_grades_m)
test_df_p = df_p.assign(grades = sum_grades_p)

In [127]:
#For our test we need to create a model where we will specify our variables 
model_m = ols ('grades ~ C(romantic) + C(sex)', data = test_df_m).fit()
model_p = ols ('grades ~ C(romantic) + C(sex)', data = test_df_p).fit()

In [135]:
#Time to calculate and see our resutls for our two-way ANOVA TEST 
anova_m = sm.stats.anova_lm(model_m, typ=2)
anova_p = sm.stats.anova_lm(model_p, typ=2)

print('Two-way ANOVA TEST results for Math class:')
print(round(anova_m, 4))
print(' ')
print('Two-way ANOVA TEST results for Portugeese class')
print(round(anova_p, 4))

Two-way ANOVA TEST results for Math class:
                 sum_sq     df       F  PR(>F)
C(romantic)    418.2249    1.0  3.4481  0.0641
C(sex)         402.3291    1.0  3.3170  0.0693
Residual     47546.6656  392.0     NaN     NaN
 
Two-way ANOVA TEST results for Portugeese class
                 sum_sq     df        F  PR(>F)
C(romantic)    524.9707    1.0   7.4308  0.0066
C(sex)         782.4584    1.0  11.0755  0.0009
Residual     45638.3238  646.0      NaN     NaN


Let's make sense of the results for our two-way ANOVA test

For Math Class
    Our P values are:
        Romance and Grades: we FAIL TO REJECT the H0 because p-value is 0.0641 and our alpha score is 0.05. So p-value > a. 
            Therfoe we can conlcude: 
                There is NO main effect of relationship status on the cumulative grades of the students. 
        Gender and grades: we FAIL TO REJECT the H0 because 
        p-value is 0.0693 and our alpha score is 0.05. So p-value > a. 
            Therfoe we can conlcude: 
                There is NO main effect of gender on the cumulative grades of the students.
    Conclusion: There is NO main effect of relationship status and gender on the cumulative grades of the students in Math Class. 

For Portugeese class
    Our P values are:
        Romance and Grades: we REJECT the H0 because p-value is 0.0066 and our alpha score is 0.05. So p-value < a. 
            Therfoe we can conlcude: 
                There is a main effect of relationship status on the cumulative grades of the students. 
        Gender and grades: we REJECT the H0 because p-value is 0.0009 and our alpha score is 0.05. So p-value < a. 
            Therfoe we can conlcude: 
                There is a main effect of gender on the cumulative grades of the students.
    Conclusion: There is a main effect of relationship status and gender on the cumulative grades of the students in Portugees class. 

Next steps
    For Math class: 
        What if see the effect of being in a relationship to the grades of studetnts after the first semester? Curriculum for math classes escalete it's difficulties after the first introductory semester. 
                H0: There is NO main effect of relationship status and gender on the cumulative grades of the students in Math class after the first semester. 
                H1: There is main effect of relationship status and gender on the cumulative grades of the students in Math class after the first semester.

In [137]:
#Step1. We need to create the sum of grades across last 2 semester for Math class
sum_grades_m_2 = df_m['G2'] + df_m['G3']

In [138]:
#Step2. We need to add this column to our new DF for this test
test_df_m_2 = df_m.assign(grades = sum_grades_m_2)

In [139]:
#Step3. For our test we need to create a model where we will specify our variables 
model_m_2 = ols ('grades ~ C(romantic) + C(sex)', data = test_df_m_2).fit()

In [140]:
#Step4. Time to calculate and see our resutls for our two-way ANOVA TEST 
anova_m_2 = sm.stats.anova_lm(model_m_2, typ=2)


print('Two-way ANOVA TEST results for Math class:')
print(round(anova_m_2, 4))

Two-way ANOVA TEST results for Math class:
                 sum_sq     df       F  PR(>F)
C(romantic)    346.2624    1.0  5.3181  0.0216
C(sex)         202.4059    1.0  3.1086  0.0787
Residual     25523.3893  392.0     NaN     NaN


Making sense of the results: 
    For Math class:
        After our manipulation and eliminating the G1 grades and running two-way ANOVA test we can conclude that: 
            There is a main effect of relationship status on the cumulative grades of the students. 
                Because: p-value for relationship between romance status and grades is 0.0215, and our alpha score is 0.05. 
                    p-value < a
                        therefore we REJECT the H0 
            There is NO main effect of gender on the cumulative grades of the student. 
                Because: p-value for relationship between romance status and grades is 0.0787, and our alpha score is 0.05. 
                    p-value >  a
                        therefore we FAIL to REJECT the H0 
    Conclusion: 
        There is a main effect of relationship status on the cumulative grades of the students in Math class after the first semester. 
            